modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/Select/Select.es.js

```diff
@@ -12,449 +12,337 @@
-		let valueArray = this._getArrayValue(value);
-
-		valueArray = this._isEmptyArray(valueArray)
-			? predefinedValueArray
-			: valueArray;
-
-		valueArray = valueArray.filter((value, index) => {
-			return multiple ? true : index === 0;
-		});
+	return newValue;
+}
 
-		const emptyOption = {
-			label: Liferay.Language.get('choose-an-option'),
-			value: '',
-		};
-
-		let newOptions = [...options]
-			.map((option, index) => {
-				return {
-					...this._prepareOption(option, valueArray),
-					separator:
-						fixedOptions.length > 0 && index === options.length - 1,
-				};
-			})
-			.concat(
-				fixedOptions.map(option =>
-					this._prepareOption(option, valueArray)
-				)
-			)
-			.filter(({value}) => value !== '');
-
-		if (!multiple) {
-			newOptions = [emptyOption, ...newOptions];
-		}
-
-		return {
-			...state,
-			options: newOptions,
-			value: valueArray.filter(value =>
-				newOptions.some(option => value === option.value)
-			),
-		};
+/**
+ * Checks if the array is empty
+ * @param arr argument to be checked if it's empty or not.
+ * @returns {Boolean}
+ */
+function isArrayEmpty(arr) {
+	if (!Array.isArray(arr) || arr.length === 0) {
+		return false;
 	}
 
-	setValue(value) {
-		const newValue = [];
-
-		if (value) {
-			newValue.push(value);
-		}
-
-		return newValue;
-	}
+	return arr.some(value => value !== '') === false;
+}
 
-	_getArrayValue(value) {
-		let newValue = value || '';
+function normalizeValue({
+	multiple,
+	normalizedOptions,
+	predefinedValueArray,
+	valueArray,
+}) {
+	const assertValue = isArrayEmpty(valueArray)
+		? predefinedValueArray
+		: valueArray;
+
+	const valueWithoutMultiple = assertValue.filter((_, index) => {
+		return multiple ? true : index === 0;
+	});
+
+	return valueWithoutMultiple.filter(value =>
+		normalizedOptions.some(option => value === option.value)
+	);
+}
 
-		if (!Array.isArray(newValue)) {
-			newValue = [newValue];
-		}
+/**
+ * Some parameters on each option
+ * needs to be prepared in case of
+ * multiple selected values(when the value state is an array).
+ *
+ */
+function assertOptionParameters({multiple, option, valueArray}) {
+	const included = valueArray.includes(option.value);
+
+	return {
+		...option,
+		active: !multiple && included,
+		checked: multiple && included,
+		type: multiple ? 'checkbox' : 'item',
+	};
+}
 
-		return newValue;
-	}
+function normalizeOptions({fixedOptions, multiple, options, valueArray}) {
+	const emptyOption = {
+		label: Liferay.Language.get('choose-an-option'),
+		value: '',
+	};
+
+	const newOptions = [...options]
+		.map((option, index) => {
+			return {
+				...assertOptionParameters({multiple, option, valueArray}),
+				separator:
+					Array.isArray(fixedOptions) &&
+					fixedOptions.length > 0 &&
+					index === options.length - 1,
+			};
+		})
+		.concat(
+			fixedOptions.map(option =>
+				assertOptionParameters({multiple, option, valueArray})
+			)
+		)
+		.filter(({value}) => value !== '');
```

Comment:
this could look like this:

```js
const newOptions = [
	...options.map((option, index) => ({
		...assertOptionParameters({multiple, option, valueArray}),
		separator:
			Array.isArray(fixedOptions) &&
			fixedOptions.length > 0 &&
			index === options.length - 1,
	})),
	...fixedOptions.map(option =>
		assertOptionParameters({multiple, option, valueArray})
	)
].filter(({value}) => value !== '');
```

Reply:
✅